langchain/src/vectorstores/typeorm.ts

```diff
@@ -0,0 +1,322 @@
+  getSourceTypeFromDocument,
+  getUniqueIDFromDocument,
+} from "../util/document_utils.js";
+import { getEnv } from "../util/env.js";
+
+export interface TypeORMVectorStoreArgs {
+  postgresConnectionOptions: DataSourceOptions;
+  tableName?: string;
+  filter?: Metadata;
+  verbose?: boolean;
+}
+
+export class TypeORMVectorStoreDocument extends Document {
+  embedding: string;
+}
+
+const defaultDocumentTableName = "documents";
+
+export class TypeORMVectorStore extends VectorStore {
+  declare FilterType: Metadata;
+
+  tableName: string;
+
+  documentEntity: EntitySchema;
+
+  filter?: Metadata;
+
+  appDataSource: DataSource;
+
+  _verbose?: boolean;
+
+  private constructor(embeddings: Embeddings, fields: TypeORMVectorStoreArgs) {
+    super(embeddings, fields);
+    this.tableName = fields.tableName || defaultDocumentTableName;
+    this.filter = fields.filter;
+  }
+
+  static async fromDataSource(
+    embeddings: Embeddings,
+    fields: TypeORMVectorStoreArgs
+  ): Promise<TypeORMVectorStore> {
+    const TypeORMDocumentEntity = new EntitySchema<TypeORMVectorStoreDocument>({
+      name: fields.tableName ?? defaultDocumentTableName,
+      columns: {
+        id: {
+          type: String,
+          primary: true,
+        },
+        pageContent: {
+          type: String,
+        },
+        metadata: {
+          type: "jsonb",
+        },
+        sourceType: {
+          type: String,
+        },
+        sourceName: {
+          type: String,
+        },
+        hash: {
+          type: String,
+        },
+        embedding: {
+          type: String,
+        },
+      },
+    });
+
+    const appDataSource = new DataSource({
+      entities: [TypeORMDocumentEntity],
+      ...fields.postgresConnectionOptions,
+    });
+
+    const postgresqlVectorStore = new TypeORMVectorStore(embeddings, fields);
+    postgresqlVectorStore.appDataSource = appDataSource;
+    postgresqlVectorStore.documentEntity = TypeORMDocumentEntity;
+
+    if (!postgresqlVectorStore.appDataSource.isInitialized) {
+      await postgresqlVectorStore.appDataSource.initialize();
+    }
+
+    postgresqlVectorStore._verbose =
+      typeof process !== "undefined"
+        ? // eslint-disable-next-line no-process-env
+          process.env?.LANGCHAIN_VERBOSE !== undefined
+        : fields?.verbose ?? false;
+
+    return postgresqlVectorStore;
+  }
+
+  async addDocuments(documents: Document[]): Promise<void> {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    // This will create the table if it does not exist. We can call it every time as it doesn't
+    // do anything if the table already exists, and it is not expensive in terms of performance
+    await this.initTable();
+    return this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents
+    );
+  }
+
+  async addVectors(vectors: number[][], documents: Document[]): Promise<void> {
+    const { createHash } = await TypeORMVectorStore.imports();
```

Comment:
We don't need to do this, just put it at the top.

Reply:
ok